### PR TITLE
pacific:mon/Elector: notify_rank_removed erase rank from both live_pinging and dead_pinging sets for highest ranked MON

### PR DESCRIPTION
**Pull request by kamoltat**
Added a case where we are removing the highest rank monitor
in `notify_rank_removed`, the old version did not deal with this
since it would only go into the loop when rank_removed < paxos_size().
Therefore, we added an else case for when rank_removed == paxos_size(),
we erase the rank from both `live_pinging` and `dead_pinging` set.

Fixes: https://tracker.ceph.com/issues/55435

Backporting relevant commit(s) from the main PR(s):

https://github.com/ceph/ceph/pull/46029

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mon/Elector.cc [hunk 1]
```diff
@@ -713,8 +713,10 @@ void Elector::notify_rank_removed(int rank_removed)
   peer_tracker.notify_rank_removed(rank_removed);
   /* we have to clean up the pinging state, which is annoying
      because it's not indexed anywhere (and adding indexing
-     would also be annoying). So what we do is start with the
-     remoed rank and examine the state of the surrounding ranks.
+     would also be annoying).
+     In the case where we are removing any rank that is not the
+     higest, we start with the removed rank and examine the state
```
Comment (ronen-fr) [inline src/mon/Elector.cc:718]: highest

File: src/mon/Elector.cc [hunk 2]
```diff
@@ -725,35 +727,46 @@ void Elector::notify_rank_removed(int rank_removed)
...
+     In the case where we are removing the highest rank,
+     we erase the removed rank from all sets.
    */
-  for (unsigned i = rank_removed + 1; i <= paxos_size() ; ++i) {
-    if (live_pinging.count(i)) {
-      dead_pinging.erase(i-1);
-      if (!live_pinging.count(i-1)) {
-	begin_peer_ping(i-1);
+  if (rank_removed < paxos_size()) {
+    for (unsigned i = rank_removed + 1; i <= paxos_size() ; ++i) {
```
Comment (ronen-fr) [inline src/mon/Elector.cc:734]: why 'unsigned' here? it forces us to cast rank_removed below

**Comment by ljflores** [comment]
Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri2-testing-2022-07-15-0755-pacific

Failures, unrelated:
    1. https://tracker.ceph.com/issues/52124
    2. https://tracker.ceph.com/issues/53294
    3. https://tracker.ceph.com/issues/53501
    4. https://tracker.ceph.com/issues/55347
    5. https://tracker.ceph.com/issues/53827
    6. https://tracker.ceph.com/issues/49287
    7. https://tracker.ceph.com/issues/53939
    8. https://tracker.ceph.com/issues/43584
    9. https://tracker.ceph.com/issues/52321
    10. https://tracker.ceph.com/issues/56652 -- new Tracker; unrelated to PRs in this run
    11. https://tracker.ceph.com/issues/49754
    12. https://tracker.ceph.com/issues/55809
    13. https://tracker.ceph.com/issues/54992
    14. https://tracker.ceph.com/issues/54071
    15. https://tracker.ceph.com/issues/56573

Details:
    1. Invalid read of size 8 in handle_recovery_delete() - Ceph - RADOS
    2. rados/test.sh hangs while running LibRadosTwoPoolsPP.TierFlushDuringFlush - Ceph - RADOS
    3. Exception when running 'rook' task. - Ceph - Orchestrator
    4. SELinux Denials during cephadm/workunits/test_cephadm - Ceph - Orchestrator
    5. cephadm exited with error code when creating osd: Input/Output error. Faulty NVME? - Ceph - Orchestrator
    6. podman: setting cgroup config for procHooks process caused: Unit libpod-$hash.scope not found - Ceph - Orchestrator
    7. ceph-nfs-upgrade, pacific: Upgrade Paused due to UPGRADE_REDEPLOY_DAEMON: Upgrading daemon osd.0 on host smithi103 failed - Ceph - Orchestrator
    8. MON_DOWN during mon_join process - Ceph - RADOS -- pending pacific backport
    9. qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Rook
    10. pacific: cephadm/test_repos.sh: rllib.error.HTTPError: HTTP Error 504: Gateway Timeout - Ceph - Orchestrator
    11. osd/OSD.cc: ceph_abort_msg("abort() called") during OSD::shutdown() - Ceph - RADOS
    12. "Leak_IndirectlyLost" valgrind report on mon.c - Ceph - RADOS
    13. cannot stat '/etc/containers/registries.conf': No such file or directory - Ceph - Orchestrator
    14. rados/cephadm/osds: Invalid command: missing required parameter hostname(<string>) - Ceph - Orchestrator
    15. test_cephadm.sh: KeyError: 'TYPE' - Ceph - Orchestrator

**Comment by ronen-fr** [comment]
@kamoltat : I know this is a backport. Can my comments be fixed without too much hassle? if not - I'll approve as-is.

File: src/mon/Elector.cc (hunk 1 above)
Comment (kamoltat) [reply src/mon/Elector.cc:718]: agreed

File: src/mon/Elector.cc (hunk 2 above)
Comment (kamoltat) [reply src/mon/Elector.cc:734]: Agreed this should be refactored

**Comment by kamoltat** [comment]
> @kamoltat : I know this is a backport. Can my comments be fixed without too much hassle? if not - I'll approve as-is.

Hi @ronen-fr thanks for the comment and yes I agree with them. However, to avoid messing up the flow of backporting, I will file another PR to main that will address all the changes.

**Comment by ronen-fr** [comment]
> Hi @ronen-fr thanks for the comment and yes I agree with them. However, to avoid messing up the flow of backporting, I will file another PR to main that will address all the changes.

Thanks

**Review by ronen-fr — APPROVED** [review]
Some minor issues to be addressed separately

**Pull request merged**